engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -744,14 +744,17 @@ protected boolean checkWorkItems(final VMInstanceVO vm, final State state) throw
 
     protected <T extends VMInstanceVO> boolean changeState(final T vm, final Event event, final Long hostId, final ItWorkVO work, final Step step) throws NoTransitionException {
         // FIXME: We should do this better.
-        final Step previousStep = work.getStep();
-        _workDao.updateStep(work, step);
+        Step previousStep = null;
+        if (work != null) {
+            previousStep = work.getStep();
```

Comment:
I know you did not write this code, but it seemed a good opportunity to discuss and evaluate it. 

I understood you. I already noticed the try/finally block, and this is the point I wanted to discuss. As the example you described, if an exception happens, the finally block is executed and the state is restored to a previous one (assuming that the `stateTransitTo(vm, event, hostId)` will change the step); and this makes sense in the case of an exception. However, if `NO` exception happens, the step is also reverted to a previous one (assuming that the `stateTransitTo(vm, event, hostId)` will change the step) . The `finally ` is always executed; either with successful or unsuccessful execution of `stateTransitTo(vm, event, hostId)`.

If we wanted to deal with exceptions, it would make much more sense executing the revert on a `catch` block. I think that we want/need to change the step for `null` when `stateTransitTo` return false and the `previousStep` is null . You are changing exactly that with the extra condition at line 757. 

Did you understand what I mean?